NuGet API v3

JoC0de opened the pull request:
This PR adds support for NuGet repositories that are using the [API version 3](https://learn.microsoft.com/en-us/nuget/api/overview). This involves some big changes. e.g.:

- Added `INugetPackage` interface with 3 different implementations for v2, v3 or local NuGet package sources.
- Version Numbers are now normalized according the [Microsoft documentation](https://learn.microsoft.com/en-us/nuget/concepts/package-versioning#where-nugetversion-diverges-from-semantic-versioning)

igor84 reviewed (approved):
Nice work on this big feature. Can't wait to try it out :)

popara96 commented on src/NuGetForUnity/Editor/NugetConfigFile.cs:
Do we maybe want to change the default package source when creating `NuGet.config` file to be V3 instead of V2?

JoC0de replied:
I changed it.

JoC0de commented on src/NuGetForUnity/Editor/NugetApiClientV3.cs:
```diff
@@ -0,0 +1,798 @@
+            // without a version specified, the latest version is returned
+            var getLatestVersion = string.IsNullOrEmpty(package.Version);
+            var item = getLatestVersion ?
+                registrationResponse.items.OrderByDescending(registrationItem => new NugetPackageVersion(registrationItem.lower)).First() :
```
I need to select the item with the highest version. With `Max(registrationItem => new NugetPackageVersion(registrationItem.lower))` I can get the latest version but not the item. Newer .Net version have the `MaxBy` that can be used for this.

JoC0de commented on src/NuGetForUnity/Editor/NugetApiClientV3.cs:
```diff
@@ -0,0 +1,798 @@
+            var getLatestVersion = string.IsNullOrEmpty(package.Version);
+            var item = getLatestVersion ?
+                registrationResponse.items.OrderByDescending(registrationItem => new NugetPackageVersion(registrationItem.lower)).First() :
+                registrationResponse.items.FirstOrDefault(
```
Until shortly I was a fan of preferring the Linq methods over the List ones to there is no mix with Linq and non Linq code. But now that even analyzers recommend using the List versions, because of the performance difference I switched.

JoC0de commented on src/NuGetForUnity/Editor/NugetApiClientV3.cs:
```diff
@@ -0,0 +1,798 @@
+            {
+                AddHeadersToRequest(request, packageSource);
+
+                using (var response = await httpClient.GetAsync(url, cancellationToken).ConfigureAwait(false))
```
Oh that was a error. I somehow overlooked it.

JoC0de commented on src/NuGetForUnity/Editor/NugetHelper.cs:
```diff
@@ -712,9 +722,10 @@ private static void DeleteAllFiles(string directoryPath, string extension)
 
         /// <summary>
-        ///     Uninstalls all of the currently installed packages.
+        ///     Uninstall's all of the currently installed packages.
```
Oh you are right. Visual Studio complaint so I fixed it (wrongly).

JoC0de commented on src/NuGetForUnity/Editor/NugetHelper.cs:
```diff
@@ -1024,7 +1032,7 @@ private static NugetPackage GetInstalledPackage(NugetPackageIdentifier packageId
                     {
                         var configPackage = PackagesConfigFile.Packages.Find(p => p.Id == packageId.Id);
-                        if (configPackage != null && configPackage < installedPackage)
+                        if (configPackage != null && configPackage.CompareVersion(installedPackage) < 0)
```
I replaced it with `configPackage.PackageVersion < installedPackage.PackageVersion`. Version is the string based version number comparing by string is not always correct.

JoC0de commented on src/NuGetForUnity/Editor/NugetHelper.cs:
```diff
@@ -1136,32 +1114,16 @@ private static NugetPackage GetOnlinePackage(NugetPackageIdentifier packageId)
-            bool isUpdate = false,
-            bool installDependencies = true)
+        /// <returns>True if the package was installed successfully, otherwise false.</returns>
+        internal static bool InstallIdentifier(INugetPackageIdentifier package, bool refreshAssets = true, bool installDependencies = true)
```
The `AlreadyImportedLibs` managed by `UnityPreImportedLibraryResolver` should never contain any NuGet package installed by NuGetForUnity, if so there are many possible issues e.g. in the UI the "update" button will not be displayed. 
It explicitly removes them, see:
https://github.com/GlitchEnzo/NuGetForUnity/blob/a37c7b09b4d2bddfc2820e4b8aa461dfe0c9d28d/src/NuGetForUnity/Editor/UnityPreImportedLibraryResolver.cs#L29-L41
Before this PR there was a edge-case that when the `NuGet.config` file was not loaded the DLL's are not excluded. I resolved it. I will also add a `if (IsAlreadyImportedInEngine(package, false))` in the start of `Uninstall` so even if the package is uninstalled before the list is filled Unity can't be tricked. The `isUpdate` solution wouldn't work in all cases e.g. when a dependency also need to be updated.

igor84 replied:
I am not sure you understood me. I am aware of the change you linked above. That code is removing things that are installed by NugetForUnity from the list of all loaded assemblies, but the issue is that it is looking at the file system to see what did NugetForUnity install and in the middle of the update procedure the package will still be in the list of all loaded assemblies and it will not be on the file system so that code you linked will not remove it from the list. Please read my first message carefully again to realize how that happens.

I don’t see how adding that check to Uninstall will help with this issue.

JoC0de replied:
The "already imported list" is static, and only filled once.
https://github.com/GlitchEnzo/NuGetForUnity/blob/a37c7b09b4d2bddfc2820e4b8aa461dfe0c9d28d/src/NuGetForUnity/Editor/UnityPreImportedLibraryResolver.cs#L14
So as long as unity doesn't reload the whole AppDomain there is no way for the package to "get into the list" while a package update is performed.

igor84 replied:
Ah, I see. Nice solution then 😀.

JoC0de commented on src/NuGetForUnity/Editor/NugetWindow.cs:
```diff
@@ -396,12 +422,18 @@ private void Refresh(bool forceFullRefresh)
+
+            // we just block the main thread
+            availablePackages = Task.Run(() => NugetHelper.Search(searchTerm, showAllOnlineVersions, showOnlinePrerelease, numberToGet, numberToSkip))
+                .Result;
```
Thanks, I fixed it.

JoC0de commented on src/NuGetForUnity/Editor/NuspecEditor.cs:
```diff
@@ -245,13 +245,6 @@ protected void OnGUI()
                         dependency.Id = EditorGUILayout.TextField(new GUIContent("ID", "The ID of the dependency package."), dependency.Id);
                         EditorGUILayout.EndHorizontal();
 
-                        //int oldSeletedIndex = IndexOf(ref existingComponents, dependency.Id);
-                        //int newSelectIndex = EditorGUILayout.Popup("Name", oldSeletedIndex, existingComponents);
-                        //if (oldSeletedIndex != newSelectIndex)
-                        //{
-                        //    dependency.Name = existingComponents[newSelectIndex];
-                        //}
-
```
It is code that is commented out since 8 years 😃 and I don't know what it should do so I think we can safely remove it 😄

JoC0de commented:
@popara96 @igor84 
I changed the way [NugetPackageVersion.cs](https://github.com/GlitchEnzo/NuGetForUnity/pull/549/files#diff-b32394e5bfda40dab7ff2e201388e17dc9b6c149370923c5f0539b8841376f16) normalizes versions because version number comparison didn't comply to the [Microsoft documentation](https://learn.microsoft.com/en-us/nuget/concepts/package-versioning).
In the implementation before the version `4.39.6.0` was not treated equal to  `4.39.6` when calling `Equals` or `GetHashCode`, only `Compare` had the correct behavior. To achieve this I changed `NugetPackageVersion.NormalizedVersion` to not contain the additional `0` at the end. So `Equals` can just use the `NormalizedVersion`.
As `INugetPackage.Version` returns the `NormalizedVersion` this has the side effect that e.g. the name of the output folder where the Package content is stored changes. Do you think this is a problem?
I am not sure if Microsoft already removes the `.0` in the API response but at least in the UI they removed it e.g. https://www.nuget.org/packages/StyleCop.MSBuild/4.7.49#versions-body-tab has no `.0`.

igor84 commented:
I never encountered a version like that (that I am aware of) so I don't see any issue with this change.

popara96 commented on src/NuGetForUnity/Editor/NugetPackageVersion.cs:
```diff
@@ -0,0 +1,602 @@
+                            var patchComparison = patchNumber.CompareTo(otherPatch);
+                            if (patchComparison == 0)
+                            {
+                                // if patch versions are equal, compare build versions
```
```suggestion
                                // if patch versions are equal, compare revision versions
```
We might want to change comments as wording has changed and it might be confusing.

igor84 commented on src/NuGetForUnity/Editor/NugetPackageV2.cs:
```diff
@@ -0,0 +1,57 @@
+        public override INugetPackageSource PackageSource => packageSourceV2;
+
+        /// <inheritdoc />
+        public override List<NugetPackageVersion> Versions => null;
```
I just noticed that this returns null but on line 40 we are passing NugetPackageV2 to FillFromNuspec method that has this line: `package.Versions.Add(package.PackageVersion);` which will cause NullReferenceException, right?

JoC0de replied:
you are right. It will be fixed in #568
